Sometimes delegate to SourceLoader in ValueSourceReaderOperator for required stored fields

martijnvg opened the pull request:
If source is required by a block loader then the `StoredFieldsSpec` that gets populated should be enhanced by `SourceLoader#requiredStoredFields(...)` in `ValuesSourceReaderOperator`. Otherwise in case of synthetic source many stored fields aren't loaded, which causes only a subset of _source to be synthesized.  For example when unmapped fields exist or field values that exceed configured ignore above will not appear is _source.

This happens when field types fallback to a block loader implementation that uses _source. The required field values are then extracted from the source once loaded.  

This change also reverts the production code changes introduced via #114903. That change only ensured that `_ignored_source` field was added to the required list of stored fields. In reality more fields could be required. This change is better fix, since it handles also other cases and the `SourceLoader` implementation indicates which stored fields are needed.

Closes #115076, #115074, #115073 and #115066

martijnvg committed:
Synthetic source and ignore above in source block loaders.

Handle ignore_above in source BlockLoader correctly when synthetic source is enabled.
This done by loading that *.original stored field that keyword fields introduce if a value is longer than specified ignore_above threshold.

Closes #115076

martijnvg committed:
simplify

martijnvg committed:
Revert "simplify"

This reverts commit 383d02ea6f168c273b0c739278caf1ed50c8b0ab.

martijnvg committed:
added yaml test

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

martijnvg committed:
Merge remote-tracking branch 'es/main' into block_source_loader_ignore_above

lkts commented on server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java:
```diff
@@ -1025,7 +1025,14 @@ protected String delegatingTo() {
+            if (sourceMode == SourceFieldMapper.Mode.SYNTHETIC && syntheticSourceDelegate().ignoreAbove() != Integer.MAX_VALUE) {
+                // Can't just read the original field, because only when ignore_above threshold is exceeded,
+                // then this stored fields gets added. In other cases this field does not exist. But we do need to include the field name:
+                String originalName = syntheticSourceDelegate().name() + "._original";
```
Can we encapsulate this logic inside `KeywordFieldMapper` somehow similar to how `syntheticSourceDelegate()` works? It would eliminate this kind of problem.

lkts commented on x-pack/plugin/logsdb/src/yamlRestTest/resources/rest-api-spec/test/50_esql_synthetic_source_disabled_fields.yml:
```diff
@@ -303,3 +303,113 @@ teardown:
+  - match: {values.0.4: false}
+  - match: {values.0.5: "POINT (-74.006 40.7128)"}
+  - match: {values.0.6: "Do. Or do not. There is no try."}
+#  - match: {values.0.7: "Do. Or do not. There is no try."}
```
Why is this commented?

martijnvg replied:
Whoops and I think this may need some extra work.

martijnvg committed:
iter

dnhatn commented:
It seems that the BlockReader for a text field with stored=false and no subfields does not return values. Should we address this issue in this PR as well?

martijnvg commented:
> It seems that the BlockReader for a text field with stored=false and no subfields does not return values. Should we address this issue in this PR as well?

Yes, this is a separate issue. The `TextFieldMapper#parseCreateField(...)` ignores that value when a text field isn't stored or doesn't have a suitable keyword subfield. I think the value should be added to ignored source if that is the case. This should be addressed on a follow up. This is also an issue for get and search apis.

martijnvg committed:
delegate sub field stuff

martijnvg committed:
Undo changes in #114903 and alter ValuesSourceReaderOperator to fully support synthetic source.

martijnvg committed:
update TODO

martijnvg committed:
Merge remote-tracking branch 'es/main' into block_source_loader_ignore_above

martijnvg commented on server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java:
```diff
@@ -1007,20 +1007,8 @@ protected String delegatingTo() {
             if (isStored()) {
                 return new BlockStoredFieldsReader.BytesFromStringsBlockLoader(name());
             }
-            if (isSyntheticSource && syntheticSourceDelegate == null) {
```
In this case the value of text fields is captured in ignored source.

martijnvg committed:
spotless

martijnvg committed:
Merge remote-tracking branch 'es/main' into block_source_loader_ignore_above

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:
```diff
@@ -241,6 +241,12 @@ private void loadFromSingleLeaf(Block[] blocks, int shard, int segment, BlockLoa
+            SourceLoader sourceLoader = null;
+            if (storedFieldsSpec.requiresSource()) {
+                sourceLoader = shardContexts.get(shard).newSourceLoader.get();
+                storedFieldsSpec = storedFieldsSpec.merge(new StoredFieldsSpec(true, true, sourceLoader.requiredStoredFields()));
```
The second parameter, `requiresMetadata`, should be false.

martijnvg replied:
Maybe it should be sourced from `storedFieldsSpec` variable (which is always initialized and gets updates by block loaders)? It could be that a block loader required metadata?

martijnvg replied:
This is what I meant: e2c77e11120e284c374490846671423507c91b16

dnhatn replied:
Yes, that works too. However, I prefer passing `false` for both the first and second arguments, as it yields the same spec since they are XORed in the `merge` method. I think it's clearer to indicate that we are appending `required stored fields` from the source loader to the spec. But I'm okay with your approach as well - just a minor nit.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:
```diff
@@ -381,9 +387,14 @@ private void fieldsMoved(LeafReaderContext ctx, int shard) throws IOException {
                 FieldWork field = fields[f];
                 rowStride[f] = field.rowStride(ctx);
                 storedFieldsSpec = storedFieldsSpec.merge(field.loader.rowStrideStoredFieldSpec());
+                SourceLoader sourceLoader = null;
```
I think the code from here to the end of the loop can be moved outside and executed once.

```java
            if (false == storedFieldsSpec.equals(StoredFieldsSpec.NO_REQUIREMENTS)) {
                trackStoredFields(storedFieldsSpec, false);
                SourceLoader sourceLoader = null;
                if (storedFieldsSpec.requiresSource()) {
                    sourceLoader = shardContexts.get(shard).newSourceLoader.get();
                    storedFieldsSpec = storedFieldsSpec.merge(new StoredFieldsSpec(true, false, sourceLoader.requiredStoredFields()));
                }
                storedFields = new BlockLoaderStoredFieldsFromLeafLoader(
                    StoredFieldLoader.fromSpec(storedFieldsSpec).getLoader(ctx, null),
                    sourceLoader != null ? sourceLoader.leaf(ctx.reader(), null) : null
                );
            }
```

martijnvg replied:
good catch: 455ecd03a692119e10f244022f5c1d8f8559ae90

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:
```diff
@@ -381,9 +387,14 @@ private void fieldsMoved(LeafReaderContext ctx, int shard) throws IOException {
+                SourceLoader sourceLoader = null;
+                if (storedFieldsSpec.requiresSource()) {
+                    sourceLoader = shardContexts.get(shard).newSourceLoader.get();
+                    storedFieldsSpec = storedFieldsSpec.merge(new StoredFieldsSpec(true, true, sourceLoader.requiredStoredFields()));
```
The second parameter, requiresMetadata, should be false.

dnhatn reviewed (commented):
I have two comments, but this approach looks better.

martijnvg committed:
properly set require source and metadata

martijnvg committed:
only load meta stored fields when required

martijnvg committed:
load sourceLoader and BlockLoaderStoredFieldsFromLeafLoader once

martijnvg committed:
Merge remote-tracking branch 'es/main' into block_source_loader_ignore_above

martijnvg committed:
fixed compile error after merging in main.

martijnvg committed:
text based block loaders don't return null loaders anymore.
and skip testBlockLoaderFromRowStrideReaderWithSyntheticSource()
if ignore above is used.

martijnvg commented on test/framework/src/main/java/org/elasticsearch/index/mapper/MapperTestCase.java:
```diff
@@ -1362,9 +1364,19 @@ public FieldNamesFieldMapper.FieldNamesFieldType fieldNames() {
+            // (both if field has doc values or stored field). While stored source just reads original values (from _source) and there
+            // is no notion of values that are ignored.
+            // TODO: fix this by improving block loader support: https://github.com/elastic/elasticsearch/issues/115257
+            assumeTrue("inconsistent synthetic source testing support with ignore above", syntheticSourceSupport.ignoreAbove() == false);
```
I'm not too happy about this. Altering the synthetic source example so that the right expected result is produced caused other test failures when stored source is used. The fact that `SyntheticSourceExample` is used for both stored and synthetic source doesn't seem right. I think the right fix is to improve the block loader testing by replacing synthetic source support by something else, like is hinted in#115257.

dnhatn reviewed (approved):
LGTM. Thanks @martijnvg

kkrik-es commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValuesSourceReaderOperator.java:
```diff
@@ -241,6 +241,14 @@ private void loadFromSingleLeaf(Block[] blocks, int shard, int segment, BlockLoa
+                storedFieldsSpec = storedFieldsSpec.merge(
+                    new StoredFieldsSpec(true, storedFieldsSpec.requiresMetadata(), sourceLoader.requiredStoredFields())
+                );
+            }
```
This seems duplicated? I wonder if we can add a helper function `getSourceLoaderLeaf` to encapsulate this logic, along with getting the leaf.

martijnvg replied:
It is duplicated, but if this is turned into a helper method it isn't necessarily getting cleaner. It would need to return a source loader and stored field spec, which means a tuple or something like that needs to be used, which makes using the helper method verbose.

lkts reviewed (commented):
I have to admit i don't really understand what's going on. Based on tests LGTM.

martijnvg committed:
Merge remote-tracking branch 'es/main' into block_source_loader_ignore_above

martijnvg committed:
use false for requiresMetadata

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 115114`
